src/read.rs

```diff
@@ -882,66 +913,60 @@ impl<R: Read + Seek> ZipArchive<R> {
             .transpose()?;
 
         #[cfg(unix)]
-        let mut files_by_unix_mode = Vec::new();
+        let mut files_by_unix_mode = FilesByUnixMode::default();
 
         for i in 0..self.len() {
             let mut file = self.by_index(i)?;
 
             let mut outpath = directory.clone();
+            /* TODO: the control flow of this method call and subsequent expectations about the
+             *       values in this loop is extremely difficult to follow. It also appears to
+             *       perform a nested loop upon extracting every single file entry? Why does it
+             *       accept two arguments that point to the same directory path, one mutable? */
             file.safe_prepare_path(directory.as_ref(), &mut outpath, root_dir.as_ref())?;
 
-            let symlink_target = if file.is_symlink() && (cfg!(unix) || cfg!(windows)) {
+            let mut symlink_target: Option<Vec<u8>> = None;
+            #[cfg(any(unix, windows))]
+            if file.is_symlink() {
                 let mut target = Vec::with_capacity(file.size() as usize);
                 file.read_to_end(&mut target)?;
-                Some(target)
-            } else {
-                if file.is_dir() {
-                    crate::read::make_writable_dir_all(&outpath)?;
-                    continue;
-                }
-                None
-            };
-
-            drop(file);
+                symlink_target = Some(target);
+            }
 
+            if symlink_target.is_none() && file.is_dir() {
+                crate::read::make_writable_dir_all(&outpath)?;
+                continue;
+            }
             if let Some(target) = symlink_target {
+                drop(file);
                 make_symlink(&outpath, &target, &self.shared.files)?;
                 continue;
             }
-            let mut file = self.by_index(i)?;
```

Comment:
This method body in general is the only place that has any non-trivial conditional logic changed. I believe the `cfg_if!` conditional expression is easier to follow.

But I particularly want to point out the one non-trivial logic change in the region I've highlighted here. The reason we needed to `drop(file)` was because of lifetime requirements for `self` on the call to `make_symlink()`. However, the previous code was unconditionally dropping the current entry handle on every single iteration.

This is very unlikely to have any noticeable effect on performance now, but it seems more correct and could be necessary for performance of future strategies. Since it's a very subtle semantic change, I wanted to call it out.